contracts/Pool.sol

```diff
@@ -14,260 +15,368 @@ import "./IBorrowersRegistry.sol";
+    uint256 residual = address(this).balance - totalSupply();
+
+    require(amount <= residual, "Trying to recover more residual funds than available");
+
+    payable(account).transfer(amount);
+  }
+
+
+  /* ========== INTERNAL FUNCTIONS ========== */
+
+  function _mint(address account, uint256 amount) internal {
+    require(account != address(0), "ERC20: mint to the zero address");
+
+    _deposited[account] += amount;
+    _deposited[address(this)] += amount;
 
-    /* ========== MODIFIERS ========== */
+    emit Transfer(address(0), account, amount);
+  }
 
-    modifier canBorrow() {
-      require(address(borrowersRegistry) != address(0), "Borrowers registry is not configured.");
-      require(borrowersRegistry.canBorrow(msg.sender), "Only the accounts authorised by borrowers registry may borrow.");
-      _;
-      require(totalDeposited > 0, "Cannot borrow from an empty pool.");
-      require(totalBorrowed * 1 ether / totalDeposited <= MAX_POOL_UTILISATION, "The pool utilisation cannot be greater than 95%.");
+
+  function _burn(address account, uint256 amount) internal {
+    require(account != address(0), "ERC20: burn from the zero address");
+
+    require(_deposited[address(this)] >= amount, "ERC20: burn amount exceeds current pool indexed balance");
+
+    require(_deposited[account] >= amount, "ERC20: burn amount exceeds user balance");
+
+    // verified in "require" above
+    unchecked {
+      _deposited[account] -= amount;
+      _deposited[address(this)] -= amount;
     }
 
+    emit Transfer(account, address(0), amount);
+  }
+
+
+  function _updateRates() internal {
+    depositIndex.setRate(_ratesCalculator.calculateDepositRate(totalBorrowed(), totalSupply()));
+    borrowIndex.setRate(_ratesCalculator.calculateBorrowingRate(totalBorrowed(), totalSupply()));
+  }
+
+
+  function _accumulateDepositInterests(address user) internal {
+    uint256 depositedWithInterests = balanceOf(user);
+    uint256 interests = depositedWithInterests - _deposited[user];
+
+    _mint(user, interests);
+
+    emit InterestsCollected(user, interests, block.timestamp);
+
+    depositIndex.updateUser(user);
+    depositIndex.updateUser(address(this));
+  }
+
+
+  function _accumulateBorrowingInterests(address user) internal {
+    uint256 borrowedWithInterests = getBorrowed(user);
+    uint256 interests = borrowedWithInterests - borrowed[user];
+    borrowed[user] = borrowedWithInterests;
+    borrowed[address(this)] += interests;
+
+    borrowIndex.updateUser(user);
+    borrowIndex.updateUser(address(this));
+  }
+
+
+  /* ========== MODIFIERS ========== */
+
+  modifier canBorrow() {
+    require(address(_borrowersRegistry) != address(0), "Borrowers registry is not configured.");
+    require(_borrowersRegistry.canBorrow(msg.sender), "Only the accounts authorised by borrowers registry may borrow.");
+    _;
+    require(totalSupply() > 0, "Cannot borrow from an empty pool.");
```

Comment:
Maybe we can move this check above the `_;` line so that in this scenario it will fail earlier and use less gass?

Reply:
right